File: collectors/snapshot/testbench/mpi_ring.c
/* MPI ring: each rank passes a token around the ring many times, plus a
 * deliberate imbalance (higher ranks sleep-spin) to exercise MPI wait time.
 * Build with mpicc. */
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

int main(int argc, char **argv) {
    MPI_Init(&argc, &argv);
    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    long rounds = (argc > 1) ? atol(argv[1]) : 20000;
    int next = (rank + 1) % size;
    int prev = (rank - 1 + size) % size;
    long token = 0;

    for (long r = 0; r < rounds; r++) {
        /* Imbalance: rank 0 does extra compute before sending. */
        if (rank == 0) {
            double acc = 1.0;
            for (int k = 0; k < 20000; k++) { acc = acc * 1.0000001 + 0.5; acc -= 0.5; }
            token += (long)acc;
        }
        if (rank == 0) {
            MPI_Send(&token, 1, MPI_LONG, next, 0, MPI_COMM_WORLD);
            MPI_Recv(&token, 1, MPI_LONG, prev, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
        } else {
            MPI_Recv(&token, 1, MPI_LONG, prev, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
            MPI_Send(&token, 1, MPI_LONG, next, 0, MPI_COMM_WORLD);
        }
    }

    MPI_Barrier(MPI_COMM_WORLD);
    if (rank == 0) printf("mpi_ring: %d ranks, %ld rounds, token=%ld\n", size, rounds, token);
    MPI_Finalize();
    return 0;
}

File: collectors/snapshot/testbench/mpi_ring.py
# mpi4py ring — Python MPI workload for validating `uaps run --mpi`.
#
# Uses stdlib `array` (no numpy needed). Uppercase Send/Recv map directly to
# the C-ABI MPI_Send / MPI_Recv, which the PMPI LD_PRELOAD shim intercepts —
# mpi4py being a runtime-imported extension does NOT prevent interception,
# because the calls bottom out in libmpi C symbols resolved via the global
# scope where LD_PRELOAD lives.
#
# Requires mpi4py built against the SAME MPI that mpirun launches (ABI match).
# Example (local venv, system OpenMPI):
#   python3 -m venv venv && MPICC=$(command -v mpicc) venv/bin/pip install mpi4py
#   uaps run --mpi -- mpirun --oversubscribe -n 4 venv/bin/python testbench/mpi_ring.py
from mpi4py import MPI
from array import array

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()
nxt = (rank + 1) % size
prv = (rank - 1 + size) % size

token = array('q', [0])
rounds = 8000
for _ in range(rounds):
    if rank == 0:
        x = 0.0
        for _ in range(2000):          # imbalance: rank 0 does extra work
            x = x * 1.0000001 + 0.5
        token[0] += 1
        comm.Send([token, MPI.LONG_LONG], dest=nxt, tag=0)
        comm.Recv([token, MPI.LONG_LONG], source=prv, tag=0)
    else:
        comm.Recv([token, MPI.LONG_LONG], source=prv, tag=0)
        comm.Send([token, MPI.LONG_LONG], dest=nxt, tag=0)

comm.Barrier()
if rank == 0:
    print(f"mpi4py ring done: {size} ranks, {rounds} rounds, token={token[0]}")
